lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -551,6 +551,25 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
     };
 
     if (!isConfused) {
+      if (Name == Context.Id_Self) {
+//        return new (Context) TypeExpr(TypeLoc::withoutLoc(
+//            DynamicSelfType::get(
+//              DC->getInnermostTypeContext()
```

Comment:
It's a little tricky. Try DC->mapTypeIntoContext(DynamicSelfType::get(DC->getInnermostTypeContext()->getSelfInterfaceType()).

Here's what's going on:

- DC->getSelfTypeInContext() ==DC->mapTypeIntoContext(DC->getSelfInterfaceType()). However, you want to map it into your DC (which might be a generic method), and not the type's DC (each DC has a new set of archetypes). We want to get rid of getSelfTypeInContext() for this reason, it's confusing.
- The difference between getDeclaredInterfaceType() and getSelfInterfaceType() is with protocols, the former returns the protocol type, whereas the latter returns the special "Self" generic parameter.

Reply:
Oh, and you only want to wrap it in a DynamicSelfType if you're inside a class -- otherwise, just getSelfInterfaceType() will do.

Reply:
Thanks for the tip @slavapestov, I've tried:
```C++
        return new (Context) TypeExpr(TypeLoc::withoutLoc(DC->
          mapTypeIntoContext(DynamicSelfType::get(DC->getInnermostTypeContext()
                                          ->getSelfInterfaceType(), Context))));
```
I'm still getting a trap that I've been seeing for a while:
```
[3/3] Linking CXX executable bin/swift
<unknown>:0: error: expected member name or constructor call after type name
Stack dump:
0.	Program arguments: /Volumes/Elements/swift-self/build/buildbot_osx/swift-macosx-x86_64/bin/swift -frontend -interpret h.swift -enable-objc-interop -sdk /Applications/Xcode1023.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -O -color-diagnostics -module-name h
1.	While type-checking 'init(a:)' (at h.swift:4:12)
2.	While type-checking statement at [h.swift:4:25 - line:7:3] RangeText="{
    print("\(Self).\(#function)")
    b = a
  "
3.	While type-checking expression at [h.swift:5:5 - line:5:33] RangeText="print("\(Self).\(#function)""
4.	While type-checking statement at [h.swift:5:11 - line:5:33] RangeText=""\(Self).\(#function)""
5.	While type-checking expression at [h.swift:5:13 - line:5:18] RangeText="(Self"
0  swift                    0x0000000110d093c5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x0000000110d08665 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x0000000110d099b8 SignalHandler(int) + 264
3  libsystem_platform.dylib 0x00007fff54501f5a _sigtramp + 26
4  swift                    0x000000011207ac38 llvm::errs()::S + 0
5  swift                    0x000000010d959cf7 diagSyntacticUseRestrictions(swift::TypeChecker&, swift::Expr const*, swift::DeclContext const*, bool)::DiagnoseWalker::checkUseOfMetaTypeName(swift::Expr*) + 583
6  swift                    0x000000010d958c9e diagSyntacticUseRestrictions(swift::TypeChecker&, swift::Expr const*, swift::DeclContext const*, bool)::DiagnoseWalker::walkToExprPre(swift::Expr*) + 2734
7  swift                    0x000000010dd3a4a9 swift::ASTVisitor<(anonymous namespace)::Traversal, swift::Expr*, swift::Stmt*, bool, swift::Pattern*, bool, void>::visit(swift::Expr*) + 201
8  swift                    0x000000010dd3ce5f (anonymous namespace)::Traversal::visitApplyExpr(swift::ApplyExpr*) + 207
9  swift                    0x000000010dd39d94 swift::Expr::walk(swift::ASTWalker&) + 84
10 swift                    0x000000010d951fe8 swift::performSyntacticExprDiagnostics(swift::TypeChecker&, swift::Expr const*, swift::DeclContext const*, bool) + 424
11 swift                    0x000000010d997f6b swift::TypeChecker::typeCheckExpressionImpl(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener&, swift::constraints::ConstraintSystem*) + 1339
12 swift                    0x000000010d997a1f swift::TypeChecker::typeCheckExpression(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener*, swift::constraints::ConstraintSystem*) + 63
13 swift                    0x000000010da2d24b swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) + 587
14 swift                    0x000000010da2c418 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) + 136
15 swift                    0x000000010da2c343 swift::TypeChecker::typeCheckTapBody(swift::TapExpr*, swift::DeclContext*) + 179
16 swift                    0x000000010d85f20f swift::constraints::ConstraintSystem::applySolution(swift::constraints::Solution&, swift::Expr*, swift::Type, bool, bool) + 735
17 swift                    0x000000010d997e47 swift::TypeChecker::typeCheckExpressionImpl(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener&, swift::constraints::ConstraintSystem*) + 1047
18 swift                    0x000000010d997a1f swift::TypeChecker::typeCheckExpression(swift::Expr*&, swift::DeclContext*, swift::TypeLoc, swift::ContextualTypePurpose, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>, swift::ExprTypeCheckListener*, swift::constraints::ConstraintSystem*) + 63
19 swift                    0x000000010da2d24b swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) + 587
20 swift                    0x000000010da2c418 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) + 136
21 swift                    0x000000010da2b392 swift::TypeChecker::typeCheckConstructorBodyUntil(swift::ConstructorDecl*, swift::SourceLoc) + 722
22 swift                    0x000000010da2badf swift::TypeChecker::typeCheckAbstractFunctionBody(swift::AbstractFunctionDecl*) + 271
23 swift                    0x000000010da4cb53 typeCheckFunctionsAndExternalDecls(swift::SourceFile&, swift::TypeChecker&) + 323
24 swift                    0x000000010da4d789 swift::performTypeChecking(swift::SourceFile&, swift::TopLevelContext&, swift::OptionSet<swift::TypeCheckingFlags, unsigned int>, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int) + 1065
25 swift                    0x000000010d182e96 swift::CompilerInstance::parseAndTypeCheckMainFileUpTo(swift::SourceFile::ASTStage_t, swift::PersistentParserState&, swift::DelayedParsingCallbacks*, swift::OptionSet<swift::TypeCheckingFlags, unsigned int>) + 502
26 swift                    0x000000010d181873 swift::CompilerInstance::parseAndCheckTypesUpTo(swift::CompilerInstance::ImplicitImports const&, swift::SourceFile::ASTStage_t) + 339
27 swift                    0x000000010d180f57 swift::CompilerInstance::performSemaUpTo(swift::SourceFile::ASTStage_t) + 615
28 swift                    0x000000010cf3df9d performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 1613
29 swift                    0x000000010cf3c8f6 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2982
30 swift                    0x000000010cee2bc2 main + 690
31 libdyld.dylib            0x00007fff541f3015 start + 1
```
In the following code:
```Swift

class A<T> {
  let b: Int
  required init(a: Int) {
    print("\(Self).\(#function)")
    b = a
  }
  static func z() {
    print("\(Self.self).\(#function)")
  }
  class func y() {
    print("\(Self).\(#function)")
    Self.z()
  }
  func x() {
    print("\(Self.self).\(#function)")
    Self.y()
    Self.z()
    _ = Self.init(a: 77)
  }
}

class B: A<Int> {
  let a: Int
  required convenience init(a: Int) {
    print("\(Self).\(#function)")
    self.init()
  }
  init() {
    print("\(Self.self).\(#function)")
    Self.y()
    Self.z()
    a = 99
    super.init(a: 88)
  }
  override class func y() {
    print("override \(Self).\(#function)")
  }
}

struct S2<T> {
  let x = 99
  static func x() {
    Self.y()
  }
  static func y() {
    print("HERE")
  }
  func foo(a: Self) -> Self {
    Self.x()
    return Self.init()
  }
}

A<Int>(a: 77).x()
B().x()
print(S2<Double>().foo(a: S2<Double>()))
//type(of:B()).init()
```

Reply:
Problem was Self without Self.self. I like that you can refer to Self on it's own like self and anyway now the error message has moved to:
```Swift
h.swift:44:10: error: generic parameter 'T' could not be inferred
    Self.y()
         ^
h.swift:50:10: error: generic parameter 'T' could not be inferred
    Self.x()
         ^
h.swift:51:17: error: generic parameter 'T' could not be inferred
    return Self.init()
```

Reply:
I've got this far but the type is still missing the generic info:
```Swift
        return new (Context) TypeExpr(TypeLoc(new (Context) FixedTypeRepr(DC->getInnermostTypeContext()->
          mapTypeIntoContext(DynamicSelfType::get(DC->getInnermostTypeContext()
                                          ->getSelfInterfaceType(), Context)), Loc)));
```
If we could solve this problem, seems like Joe's `x as? Self` would work.

Reply:
The `could not be inferred` problem above is fixed if I use the following code:
```C++
        Type SelfType = DC->getInnermostTypeContext()->getSelfInterfaceType();
        if (ClassType::classof(SelfType.getPointer()))
          SelfType = DynamicSelfType::get(SelfType, Context);
        return new (Context) TypeExpr(TypeLoc(new (Context) FixedTypeRepr(DC->getInnermostTypeContext()
              ->mapTypeIntoContext(SelfType), Loc)));
```
This leave Joe's `x as? Self` which now works now provided Self is not generic.

Reply:
`ClassType::classof(SelfType.getPointer())` should be written as `SelfType->is<ClassType>()`. However, that doesn't account for generic classes, so really you want to check that `DC->getSelfDecl()` is non-null.